lib/buffer.js

```diff
@@ -1021,557 +999,768 @@ Buffer.prototype.slice = function slice(start, end) {
-// multiplying the fraction with the result of 2**-bits_in_fraction.
-
-
-function toDouble(x0, x1) {
-  const frac = x0 + 0x100000000 * (x1 & 0xFFFFF);
-  const expt = (x1 >>> 20) & 2047;
-  const sign = (x1 >>> 31) ? -1 : 1;
-  if (expt === 0) {
-    if (frac === 0) return sign * 0;
-    return sign * frac * 2 ** -1074;
-  } else if (expt === 2047) {
-    if (frac === 0) return sign * Infinity;
-    return NaN;
-  }
-  return sign * 2 ** (expt - 1023) * (1 + frac * 2 ** -52);
+function readInt40LE(buf, offset) {
+  checkNumberType(offset);
+  const first = buf[offset];
+  const last = buf[offset + 4];
+  if (first === undefined || last === undefined)
+    boundsError(offset, buf.length - 5);
+
+  return (last < 2 ** 7 ? last : last | 0xffffff00) * 2 ** 32 +
+    first +
+    buf[++offset] * 2 ** 8 +
+    buf[++offset] * 2 ** 16 +
+    buf[++offset] * 2 ** 24;
 }
 
+Buffer.prototype.readInt32LE = function readInt32LE(offset) {
+  checkNumberType(offset);
+  const first = this[offset];
+  const last = this[offset + 3];
+  if (first === undefined || last === undefined)
+    boundsError(offset, this.length - 4);
+
+  return first +
+    this[++offset] * 2 ** 8 +
+    this[++offset] * 2 ** 16 +
+    (last << 24); // Overflow
+};
 
-function toFloat(x) {
-  const frac = x & 0x7FFFFF;
-  const expt = (x >>> 23) & 255;
-  const sign = (x >>> 31) ? -1 : 1;
-  if (expt === 0) {
-    if (frac === 0) return sign * 0;
-    return sign * frac * 2 ** -149;
-  } else if (expt === 255) {
-    if (frac === 0) return sign * Infinity;
-    return NaN;
-  }
-  return sign * 2 ** (expt - 127) * (1 + frac * 2 ** -23);
+function readInt24LE(buf, offset) {
+  checkNumberType(offset);
+  const first = buf[offset];
+  const last = buf[offset + 2];
+  if (first === undefined || last === undefined)
+    boundsError(offset, buf.length - 3);
+
+  const val = first + buf[++offset] * 2 ** 8 + last * 2 ** 16;
+  return val < 2 ** 23 ? val : val | 0xff000000;
 }
 
+Buffer.prototype.readInt16LE = function readInt16LE(offset) {
+  checkNumberType(offset);
+  const first = this[offset];
+  const last = this[offset + 1];
+  if (first === undefined || last === undefined)
+    boundsError(offset, this.length - 2);
 
-Buffer.prototype.readDoubleBE = function(offset, noAssert) {
-  offset = offset >>> 0;
-  const x1 = this.readUInt32BE(offset + 0, noAssert);
-  const x0 = this.readUInt32BE(offset + 4, noAssert);
-  return toDouble(x0, x1);
+  const val = first + last * 2 ** 8;
+  return val < 2 ** 15 ? val : val | 0xffff0000;
 };
 
+Buffer.prototype.readInt8 = function readInt8(offset) {
+  checkNumberType(offset);
+  const val = this[offset];
+  if (val === undefined)
+    boundsError(offset, this.length - 1);
 
-Buffer.prototype.readDoubleLE = function(offset, noAssert) {
-  offset = offset >>> 0;
-  const x0 = this.readUInt32LE(offset + 0, noAssert);
-  const x1 = this.readUInt32LE(offset + 4, noAssert);
-  return toDouble(x0, x1);
+  return val < 2 ** 7 ? val : val | 0xffffff00;
```

Comment:
There is a branchless alternative:
```js
return val | (val & 128) * 0x1fffffe;  // 128*0x1fffffe == 0xffffff00
```
(I'm pointing it out here but the concept can be applied to other readInt methods.)

Another take on the same principle:
```js
return val * (1 | -(val >>> 7));  // multiply by 1 or -1
```

Reply:
D'oh! Of course.

In version b) you probably meant:

```js
return val - (2 ** 8 * (val >>> 7));
```

I just benchmarked the three variations and the current implementation is the same compared the your b. Your version a) wins (3600ms vs 3600ms vs 3200ms in my benchmark). So I switched accordingly.